### PR TITLE
Optimizations for DiscreteTrajectory

### PR DESCRIPTION
**Pull request by rnlahaye**
To `front` and `end`.
Also, renamed the existing Front/Back/etc benchmarks as BM_DiscreteTrajectorySegmentFoo (because they measure the performance of segment functions), and created new benchmarks that measure the corresponding functions on the entire trajectory.

Before optimizations, but including benchmark reconfiguration (#5b6df36):
```
Benchmark                                                          Time             CPU   Iterations
----------------------------------------------------------------------------------------------------
BM_DiscreteTrajectoryFront                                      2.06 ns         2.06 ns    335252253
BM_DiscreteTrajectoryBack                                       18.5 ns         18.5 ns     37876943
BM_DiscreteTrajectoryBegin                                      1.22 ns         1.22 ns    572222676
BM_DiscreteTrajectoryEnd                                        7.84 ns         7.83 ns     89294826
BM_DiscreteTrajectoryTMin                                       2.34 ns         2.34 ns    298857936
BM_DiscreteTrajectoryTMax                                       4.41 ns         4.41 ns    158582724
BM_DiscreteTrajectorySegmentFront                              0.316 ns        0.316 ns   1000000000
BM_DiscreteTrajectorySegmentBack                               0.590 ns        0.590 ns   1000000000
BM_DiscreteTrajectorySegmentBegin                              0.789 ns        0.789 ns    884765600
BM_DiscreteTrajectorySegmentEnd                                 6.32 ns         6.31 ns    110878794
BM_DiscreteTrajectorySegmentTMin                                1.90 ns         1.90 ns    369151584
BM_DiscreteTrajectorySegmentTMax                                4.42 ns         4.42 ns    158637350
BM_DiscreteTrajectoryCreateDestroy/8                            1323 ns         1322 ns       528753
BM_DiscreteTrajectoryCreateDestroy/64                           4301 ns         4299 ns       163910
BM_DiscreteTrajectoryCreateDestroy/512                         27396 ns        27381 ns        25575
BM_DiscreteTrajectoryCreateDestroy/1024                        54314 ns        54277 ns        12885
BM_DiscreteTrajectoryIterate/8                                  21.6 ns         21.5 ns     32600900
BM_DiscreteTrajectoryIterate/64                                  152 ns          151 ns      4615801
BM_DiscreteTrajectoryIterate/512                                1180 ns         1179 ns       592934
BM_DiscreteTrajectoryIterate/1024                               2326 ns         2325 ns       299891
BM_DiscreteTrajectoryReverseIterate/8                           19.5 ns         19.5 ns     35863412
BM_DiscreteTrajectoryReverseIterate/64                           181 ns          181 ns      3881320
BM_DiscreteTrajectoryReverseIterate/512                         1490 ns         1489 ns       469358
BM_DiscreteTrajectoryReverseIterate/1024                        2940 ns         2939 ns       238825
BM_DiscreteTrajectoryFind/8                                     95.5 ns         95.4 ns      7303535
BM_DiscreteTrajectoryFind/64                                     121 ns          121 ns      5787994
BM_DiscreteTrajectoryFind/512                                    139 ns          139 ns      4949515
BM_DiscreteTrajectoryFind/1024                                   145 ns          144 ns      4830318
BM_DiscreteTrajectoryLowerBound/8                               58.5 ns         58.4 ns     11962540
BM_DiscreteTrajectoryLowerBound/64                              73.3 ns         73.2 ns      9503122
BM_DiscreteTrajectoryLowerBound/512                             90.3 ns         90.3 ns      7786256
BM_DiscreteTrajectoryLowerBound/1024                            94.1 ns         94.0 ns      7421386
BM_DiscreteTrajectoryEvaluateDegreesOfFreedomExact              8.51 ns         8.51 ns     82096992
BM_DiscreteTrajectoryEvaluateDegreesOfFreedomInterpolated       20.6 ns         20.6 ns     34142844
```

After optimizations (#dc232bd):
```
Benchmark                                                          Time             CPU   Iterations
----------------------------------------------------------------------------------------------------
BM_DiscreteTrajectoryFront                                     0.448 ns        0.448 ns   1000000000
BM_DiscreteTrajectoryBack                                       3.67 ns         3.67 ns    190162100
BM_DiscreteTrajectoryBegin                                      1.30 ns         1.30 ns    575232147
BM_DiscreteTrajectoryEnd                                       0.631 ns        0.631 ns   1000000000
BM_DiscreteTrajectoryTMin                                       2.39 ns         2.39 ns    297638030
BM_DiscreteTrajectoryTMax                                       2.63 ns         2.63 ns    269664306
BM_DiscreteTrajectorySegmentFront                              0.316 ns        0.316 ns   1000000000
BM_DiscreteTrajectorySegmentBack                               0.589 ns        0.589 ns   1000000000
BM_DiscreteTrajectorySegmentBegin                              0.789 ns        0.788 ns    886019872
BM_DiscreteTrajectorySegmentEnd                                 6.32 ns         6.32 ns    110255320
BM_DiscreteTrajectorySegmentTMin                                1.89 ns         1.89 ns    368419114
BM_DiscreteTrajectorySegmentTMax                                2.22 ns         2.22 ns    316769315
BM_DiscreteTrajectoryCreateDestroy/8                            1280 ns         1279 ns       537403
BM_DiscreteTrajectoryCreateDestroy/64                           4220 ns         4217 ns       167352
BM_DiscreteTrajectoryCreateDestroy/512                         26729 ns        26713 ns        25983
BM_DiscreteTrajectoryCreateDestroy/1024                        53650 ns        53619 ns        13030
BM_DiscreteTrajectoryIterate/8                                  20.8 ns         20.8 ns     33638159
BM_DiscreteTrajectoryIterate/64                                  143 ns          143 ns      4878627
BM_DiscreteTrajectoryIterate/512                                1201 ns         1201 ns       579005
BM_DiscreteTrajectoryIterate/1024                               2398 ns         2397 ns       291498
BM_DiscreteTrajectoryReverseIterate/8                           19.2 ns         19.2 ns     36514055
BM_DiscreteTrajectoryReverseIterate/64                           181 ns          181 ns      3841763
BM_DiscreteTrajectoryReverseIterate/512                         1492 ns         1491 ns       468751
BM_DiscreteTrajectoryReverseIterate/1024                        2943 ns         2941 ns       238100
BM_DiscreteTrajectoryFind/8                                     79.3 ns         79.2 ns      8869629
BM_DiscreteTrajectoryFind/64                                    96.3 ns         96.3 ns      7271443
BM_DiscreteTrajectoryFind/512                                    117 ns          117 ns      5980912
BM_DiscreteTrajectoryFind/1024                                   118 ns          118 ns      5954660
BM_DiscreteTrajectoryLowerBound/8                               58.9 ns         58.8 ns     11972566
BM_DiscreteTrajectoryLowerBound/64                              73.2 ns         73.1 ns      9578151
BM_DiscreteTrajectoryLowerBound/512                             87.3 ns         87.3 ns      7903889
BM_DiscreteTrajectoryLowerBound/1024                            93.8 ns         93.8 ns      7445066
BM_DiscreteTrajectoryEvaluateDegreesOfFreedomExact              8.56 ns         8.56 ns     82251337
BM_DiscreteTrajectoryEvaluateDegreesOfFreedomInterpolated       18.4 ns         18.4 ns     38056715
```

Diff:
```
Benchmark                                                                   Time             CPU      Time Old      Time New       CPU Old       CPU New
--------------------------------------------------------------------------------------------------------------------------------------------------------
BM_DiscreteTrajectoryFront                                               -0.7831         -0.7831             2             0             2             0
BM_DiscreteTrajectoryBack                                                -0.8015         -0.8015            19             4            19             4
BM_DiscreteTrajectoryBegin                                               +0.0697         +0.0695             1             1             1             1
BM_DiscreteTrajectoryEnd                                                 -0.9195         -0.9195             8             1             8             1
BM_DiscreteTrajectoryTMin                                                +0.0219         +0.0219             2             2             2             2
BM_DiscreteTrajectoryTMax                                                -0.4038         -0.4037             4             3             4             3
BM_DiscreteTrajectorySegmentFront                                        -0.0001         -0.0003             0             0             0             0
BM_DiscreteTrajectorySegmentBack                                         -0.0022         -0.0022             1             1             1             1
BM_DiscreteTrajectorySegmentBegin                                        -0.0009         -0.0011             1             1             1             1
BM_DiscreteTrajectorySegmentEnd                                          +0.0008         +0.0009             6             6             6             6
BM_DiscreteTrajectorySegmentTMin                                         -0.0022         -0.0024             2             2             2             2
BM_DiscreteTrajectorySegmentTMax                                         -0.4973         -0.4974             4             2             4             2
BM_DiscreteTrajectoryCreateDestroy/8                                     -0.0326         -0.0327          1323          1280          1322          1279
BM_DiscreteTrajectoryCreateDestroy/64                                    -0.0188         -0.0189          4301          4220          4299          4217
BM_DiscreteTrajectoryCreateDestroy/512                                   -0.0243         -0.0244         27396         26729         27381         26713
BM_DiscreteTrajectoryCreateDestroy/1024                                  -0.0122         -0.0121         54314         53650         54277         53619
BM_DiscreteTrajectoryIterate/8                                           -0.0338         -0.0335            22            21            22            21
BM_DiscreteTrajectoryIterate/64                                          -0.0559         -0.0558           152           143           151           143
BM_DiscreteTrajectoryIterate/512                                         +0.0183         +0.0182          1180          1201          1179          1201
BM_DiscreteTrajectoryIterate/1024                                        +0.0312         +0.0310          2326          2398          2325          2397
BM_DiscreteTrajectoryReverseIterate/8                                    -0.0144         -0.0145            20            19            20            19
BM_DiscreteTrajectoryReverseIterate/64                                   -0.0006         -0.0008           181           181           181           181
BM_DiscreteTrajectoryReverseIterate/512                                  +0.0009         +0.0009          1490          1492          1489          1491
BM_DiscreteTrajectoryReverseIterate/1024                                 +0.0009         +0.0007          2940          2943          2939          2941
BM_DiscreteTrajectoryFind/8                                              -0.1697         -0.1698            95            79            95            79
BM_DiscreteTrajectoryFind/64                                             -0.2051         -0.2050           121            96           121            96
BM_DiscreteTrajectoryFind/512                                            -0.1617         -0.1618           139           117           139           117
BM_DiscreteTrajectoryFind/1024                                           -0.1836         -0.1838           145           118           144           118
BM_DiscreteTrajectoryLowerBound/8                                        +0.0070         +0.0070            58            59            58            59
BM_DiscreteTrajectoryLowerBound/64                                       -0.0015         -0.0016            73            73            73            73
BM_DiscreteTrajectoryLowerBound/512                                      -0.0333         -0.0332            90            87            90            87
BM_DiscreteTrajectoryLowerBound/1024                                     -0.0028         -0.0029            94            94            94            94
BM_DiscreteTrajectoryEvaluateDegreesOfFreedomExact                       +0.0060         +0.0059             9             9             9             9
BM_DiscreteTrajectoryEvaluateDegreesOfFreedomInterpolated                -0.1061         -0.1062            21            18            21            18
```

Note that the TMax benchmarks appear to be noisy (TMax is not affected by this change).

### CODE REVIEW

File: physics/discrete_trajectory_body.hpp [hunk 1]
```diff
@@ -33,7 +33,7 @@ DiscreteTrajectory<Frame>::DiscreteTrajectory()
 template<typename Frame>
 typename DiscreteTrajectory<Frame>::reference
 DiscreteTrajectory<Frame>::front() const {
-  return *begin();
+  return segments_->front().front();
```
Comment (eggrobin) [inline physics/discrete_trajectory_body.hpp:36]: We don’t think this is correct: `segments_->front()` can be empty; the design doc is out of date, as all design docs are, but there are comments to that effect: https://github.com/mockingbirdnest/Principia/blob/d9e6442827e5006d2e700d1baf45ccef03f25a35/physics/discrete_trajectory.hpp#L99-L101 https://github.com/mockingbirdnest/Principia/blob/d9e6442827e5006d2e700d1baf45ccef03f25a35/physics/discrete_trajectory.hpp#L218-L223

Comment (eggrobin) [reply physics/discrete_trajectory_body.hpp:36]: Arguably, the fact that it is incorrect—and the fact that it is a performance improvement—seems like a bug in the implementation of `front` as changed in #3207 ; we end up with a `front()` with semantics that differ from `*begin()`, which seems distasteful.

In any case, please add a test that would have caught this.

It might be possible to actually optimize things using `segment_by_left_endpoint_`.

Comment (eggrobin) [reply physics/discrete_trajectory_body.hpp:36]: OK, having discussed this with @pleroy, I think we want the following:
- fix `front` on segments (to do `*begin` again; #3207 probably predates the existence of empty segments, so it would have been correct at the time, but `front` on a segment is very rarely used, so nobody should care about its performance anyway);
- add a test that would catch brokenness on `front` on either an empty segment or a trajectory starting with empty segments;
- use `segments_by_left_endpoint_` to write a faster `front` on trajectories (maybe you could even use it to write a faster begin, and keep `front` as `*begin`?).

Comment (rnlahaye) [reply physics/discrete_trajectory_body.hpp:36]: Update: I have a test for `front` on a trajectory starting with an empty segment, and a fast correct version of trajectory `front` using `segments_by_left_endpoint_`. I'll fix segment `front` and look into improving trajectory `begin`, as requested.

Comment (rnlahaye) [reply physics/discrete_trajectory_body.hpp]: I wasn't able to speed up `begin`.
Here are the benchmarks for the updated functions:
```
Benchmark                                  Time             CPU   Iterations
----------------------------------------------------------------------------
BM_DiscreteTrajectoryFront             0.592 ns        0.592 ns   1000000000
BM_DiscreteTrajectorySegmentFront      0.526 ns        0.525 ns   1000000000
```
```
Benchmark                                           Time             CPU      Time Old      Time New       CPU Old       CPU New
BM_DiscreteTrajectoryFront                       -0.7197         -0.7196             2             1             2             1
BM_DiscreteTrajectorySegmentFront                +0.6716         +0.6716             0             1             0             1
```

Comment (eggrobin) [reply physics/discrete_trajectory_body.hpp]: > I wasn't able to speed up `begin`.

Since you return a `DiscreteTrajectoryIterator`, and thus need to go through the constructor of `DiscreteTrajectoryIterator`, I don’t think you can make it faster in the simple case (no empty segments at the beginning), which is what you are benchmarking.

However, in हरीश चंद्र, we can reasonably expect to have hundreds of empty segments at the beginning (for instance, an old LEO sat that did occasional corrections with RCS will have two segments—burn and coast—for every RCS firing; most of those will not have been reanimated unless you looked at its entire history all the way to launch), so I suspect this may well be worth it.

Comment (rnlahaye) [reply physics/discrete_trajectory_body.hpp]: > we can reasonably expect to have hundreds of empty segments at the beginning
I'll go ahead and add another benchmark then!

Comment (rnlahaye) [reply physics/discrete_trajectory_body.hpp]: Benchmarks for trajectories beginning with 100 empty segments:
Pre-optimizations (note that this is with the buggy `Segment::front` code):
```
Benchmark                                       Time             CPU   Iterations
---------------------------------------------------------------------------------
BM_DiscreteTrajectoryFrontEmpty               113 ns          113 ns      6195622
BM_DiscreteTrajectoryBeginEmpty              88.8 ns         88.8 ns      7838307
BM_DiscreteTrajectorySegmentFrontEmpty      0.315 ns        0.315 ns   1000000000
BM_DiscreteTrajectorySegmentBeginEmpty       88.6 ns         88.6 ns      7969942
```
Post-optimizations (including the change to `DiscreteTrajectory::begin`):
```
Benchmark                                       Time             CPU   Iterations
---------------------------------------------------------------------------------
BM_DiscreteTrajectoryFrontEmpty             0.592 ns        0.592 ns   1000000000
BM_DiscreteTrajectoryBeginEmpty              1.46 ns         1.46 ns    479777383
BM_DiscreteTrajectorySegmentFrontEmpty        110 ns          110 ns      6333409
BM_DiscreteTrajectorySegmentBeginEmpty       88.5 ns         88.4 ns      7939659
```
Delta:
```
Benchmark                                                Time             CPU      Time Old      Time New       CPU Old       CPU New
-------------------------------------------------------------------------------------------------------------------------------------
BM_DiscreteTrajectoryFrontEmpty                       -0.9948         -0.9948           113             1           113             1
BM_DiscreteTrajectoryBeginEmpty                       -0.9836         -0.9836            89             1            89             1
BM_DiscreteTrajectorySegmentFrontEmpty              +349.6114       +349.5768             0           110             0           110
BM_DiscreteTrajectorySegmentBeginEmpty                -0.0016         -0.0014            89            88            89            88
```
`Segment::front` sees a massive regression (because it was broken before), while `DiscreteTrajectory::front` and `DiscreteTrajectory::begin` see massive gains.
The change to `DiscreteTrajectory::begin` causes a mild regression for the non-empty-segment benchmark:
```
Benchmark                                    Time             CPU      Time Old      Time New       CPU Old       CPU New
-------------------------------------------------------------------------------------------------------------------------
BM_DiscreteTrajectoryBegin                +0.1961         +0.1960             1             1             1             1
```
It's still under 2 ns though so I think it's ok.

**Comment by rnlahaye** [comment]
I optimized `DiscreteTrajectory::find` and friends somewhat by bypassing the call to `DiscreteTrajectorySegment::end`.

Old:
```
Benchmark                                     Time             CPU   Iterations
-------------------------------------------------------------------------------
BM_DiscreteTrajectoryFind/8                97.9 ns         97.4 ns      7185752
BM_DiscreteTrajectoryFind/64                122 ns          122 ns      5721619
BM_DiscreteTrajectoryFind/512               140 ns          139 ns      4987638
BM_DiscreteTrajectoryFind/1024              143 ns          142 ns      4952456
BM_DiscreteTrajectoryLowerBound/8          59.6 ns         59.5 ns     11828720
BM_DiscreteTrajectoryLowerBound/64         74.2 ns         74.1 ns      9457287
BM_DiscreteTrajectoryLowerBound/512        92.4 ns         92.3 ns      7593262
BM_DiscreteTrajectoryLowerBound/1024       95.1 ns         95.0 ns      7329996
```

New:
```
Benchmark                                     Time             CPU   Iterations
-------------------------------------------------------------------------------
BM_DiscreteTrajectoryFind/8                22.7 ns         22.7 ns     30914222
BM_DiscreteTrajectoryFind/64               34.4 ns         34.4 ns     20379644
BM_DiscreteTrajectoryFind/512              51.5 ns         51.4 ns     13518472
BM_DiscreteTrajectoryFind/1024             52.5 ns         52.4 ns     13293832
BM_DiscreteTrajectoryLowerBound/8          36.9 ns         36.8 ns     18994847
BM_DiscreteTrajectoryLowerBound/64         53.7 ns         53.7 ns     13234265
BM_DiscreteTrajectoryLowerBound/512        74.4 ns         74.4 ns      9413410
BM_DiscreteTrajectoryLowerBound/1024       76.3 ns         76.2 ns      9148533
```

Delta:
```
Benchmark                                              Time             CPU      Time Old      Time New       CPU Old       CPU New
-----------------------------------------------------------------------------------------------------------------------------------
BM_DiscreteTrajectoryFind/8                         -0.7685         -0.7674            98            23            97            23
BM_DiscreteTrajectoryFind/64                        -0.7181         -0.7181           122            34           122            34
BM_DiscreteTrajectoryFind/512                       -0.6311         -0.6312           140            51           139            51
BM_DiscreteTrajectoryFind/1024                      -0.6318         -0.6318           143            52           142            52
BM_DiscreteTrajectoryLowerBound/8                   -0.3814         -0.3815            60            37            60            37
BM_DiscreteTrajectoryLowerBound/64                  -0.2758         -0.2760            74            54            74            54
BM_DiscreteTrajectoryLowerBound/512                 -0.1939         -0.1942            92            74            92            74
BM_DiscreteTrajectoryLowerBound/1024                -0.1976         -0.1977            95            76            95            76
```
Note that some of the gains in `BM_DiscreteTrajectoryFind` are due to the earlier changes to `DiscreteTrajectory::end`.

File: physics/discrete_trajectory_body.hpp [hunk 2]
```diff
@@ -45,13 +45,14 @@ DiscreteTrajectory<Frame>::back() const {
...
 DiscreteTrajectory<Frame>::begin() const {
-  return segments_->front().begin();
+  return empty() ? end() : segment_by_left_endpoint_.begin()->second->begin();
 }
 
 template<typename Frame>
 typename DiscreteTrajectory<Frame>::iterator
 DiscreteTrajectory<Frame>::end() const {
-  return segments_->back().end();
+  return iterator(SegmentIterator(segments_.get(), segments_->end()),
```
Comment (pleroy) [inline physics/discrete_trajectory_body.hpp:54]: It irks me a bit that we are leaking here details about the representation of the end of the trajectory (`point_ == std::nullopt`).  A possibility would be to have a (private) factory `DiscreteTrajectoryIterator::EndOfLastSegment` with a single parameter (`segment`) which would (1) DCHECK that segment `is_end()` and (2) reset `point_`.  That way the ugly details would stay in `DiscreteTrajectoryIterator`.

Of course, this may have a performance cost, but it may not because there is a lot of inlining happening in these classes.

File: physics/discrete_trajectory_segment.hpp [hunk 3]
```diff
@@ -136,6 +136,12 @@ class DiscreteTrajectorySegment : public Trajectory<Frame> {
       DiscreteTrajectorySegmentIterator<Frame> self);
 
  private:
+  // Versions of find, lower_bound, and upper_bound that use optionals to
+  // indicate absense. Useful for bypassing expensive end() call.
+  std::optional<iterator> FindImpl(Instant const& t) const;
```
Comment (pleroy) [inline physics/discrete_trajectory_segment.hpp:141]: I don't think that we love the suffix `Impl`, we don't use that a lot.  Furthermore, it's not so clear at the call site how the `Impl` differ from the public method.  I'd call these `FindOrNullopt`, etc. which will make the role of these methods more explicit.

File: physics/discrete_trajectory_segment_body.hpp [hunk 4]
```diff
@@ -107,9 +107,16 @@ void DiscreteTrajectorySegment<Frame>::clear() {
 template<typename Frame>
 typename DiscreteTrajectorySegment<Frame>::iterator
 DiscreteTrajectorySegment<Frame>::find(Instant const& t) const {
+  auto const it = FindImpl(t);
+  return it.has_value() ? *it : end();
+}
+
+template <typename Frame>
+std::optional<typename DiscreteTrajectorySegment<Frame>::iterator>
+DiscreteTrajectorySegment<Frame>::FindImpl(Instant const& t) const {
```
Comment (pleroy) [inline physics/discrete_trajectory_segment_body.hpp:116]: We try to use the same order for the methods in the `.hpp` and `.cpp` files, so these implementation methods, which are private, should move down.

File: benchmarks/discrete_trajectory.cpp [hunk 5]
```diff
@@ -63,8 +63,8 @@ DiscreteTrajectory<World> MakeTrajectory(Timeline<World> const& timeline,
         double const split = *it_split;
         CHECK_LE(0, split);
         CHECK_LE(split, 1);
-        t_split = Barycentre<Instant, double>({t_min, t_max},
-                                              {1 - split, split});
```
Comment (pleroy) [inline benchmarks/discrete_trajectory.cpp:67]: I'd rather not change the formatting (we don't systematically use clang-format).

File: benchmarks/discrete_trajectory.cpp [hunk 6]
```diff
@@ -77,6 +77,18 @@ DiscreteTrajectory<World> MakeTrajectory(Timeline<World> const& timeline,
   return trajectory;
 }
 
+// Constructs a trajectory beginning with many empty segments.
+DiscreteTrajectory<World> MakeTrajectoryWithEmptySegments(
+    int num_empty_segments) {
```
Comment (pleroy) [inline benchmarks/discrete_trajectory.cpp:82]: `int const`

**Review by pleroy — CHANGES_REQUESTED** [review]

File: physics/discrete_trajectory_body.hpp [hunk 7]
```diff
@@ -33,7 +33,7 @@ DiscreteTrajectory<Frame>::DiscreteTrajectory()
 template<typename Frame>
 typename DiscreteTrajectory<Frame>::reference
 DiscreteTrajectory<Frame>::front() const {
-  return *begin();
+  return *segment_by_left_endpoint_.begin()->second->timeline_.begin();
```
[the remark this answers is not in the document]
Comment (rnlahaye) [reply physics/discrete_trajectory_body.hpp]: Done.

File: physics/discrete_trajectory_body.hpp (hunk 2 above)
Comment (rnlahaye) [reply physics/discrete_trajectory_body.hpp]: Done. No performance penalty.

File: physics/discrete_trajectory_segment.hpp (hunk 3 above)
Comment (rnlahaye) [reply physics/discrete_trajectory_segment.hpp]: Done.

File: physics/discrete_trajectory_segment_body.hpp (hunk 4 above)
Comment (rnlahaye) [reply physics/discrete_trajectory_segment_body.hpp]: Done.

File: benchmarks/discrete_trajectory.cpp (hunk 5 above)
Comment (rnlahaye) [reply benchmarks/discrete_trajectory.cpp]: Whoops. Fixed.

File: benchmarks/discrete_trajectory.cpp (hunk 6 above)
Comment (rnlahaye) [reply benchmarks/discrete_trajectory.cpp]: Done.

File: benchmarks/discrete_trajectory.cpp [hunk 8]
```diff
@@ -79,7 +79,7 @@ DiscreteTrajectory<World> MakeTrajectory(Timeline<World> const& timeline,
 
 // Constructs a trajectory beginning with many empty segments.
 DiscreteTrajectory<World> MakeTrajectoryWithEmptySegments(
-    int num_empty_segments) {
+    int const num_empty_segments) {
```
Comment (pleroy) [inline benchmarks/discrete_trajectory.cpp:82]: One last nit: we don't use `num_`.  We would either write `number_of_empty_segments` or just `empty_segments`.  Pick what you prefer.

Comment (rnlahaye) [reply benchmarks/discrete_trajectory.cpp]: Done.

**Review by eggrobin — APPROVED** [review]

**Comment by pleroy** [comment]
FWIW I have appended a blurb to the [design document](https://docs.google.com/document/d/1POE2fXMCMndNPzhHelkrmKAS1TvFsZ7pBOZk_9sqtJs/edit?usp=sharing) explaining why we need empty segments and how they work.
